File: 125.valid-palindrome.js
/*
 * @lc app=leetcode id=125 lang=javascript
 *
 * [125] Valid Palindrome
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  const isAlphanumeric = (ch) => {
    return ("a" <= ch && ch <= "z") || ("0" <= ch && ch <= "9");
  };
  const str = s.toLowerCase();
  for (let i = 0, j = str.length - 1; i < j; ++i, --j) {
    while (!isAlphanumeric(str[i]) && i < j) ++i;
    while (!isAlphanumeric(str[j]) && i < j) --j;
    if (i < j && str[i] !== str[j]) return false;
  }
  return true;
};
// @lc code=end

File: 125.valid-palindrome.go
/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */

package leetcode

import "strings"

// @lc code=start
func isPalindrome125(s string) bool {
	s = strings.ToLower(s)
	i, j := 0, len(s)-1
	for i < j {
		for i < j && !isChar125(s[i]) {
			i++
		}
		for i < j && !isChar125(s[j]) {
			j--
		}
		if s[i] != s[j] {
			return false
		}
		i, j = i+1, j-1
	}
	return true
}

func isChar125(c byte) bool {
	if 'a' <= c && c <= 'z' || '0' <= c && c <= '9' {
		return true
	}
	return false
}

// @lc code=end
